hastory-data/src/Data/Hastory/API.hs

```diff
@@ -1,84 +1,90 @@
+type AuthCookies = '[ Header "Set-Cookie" Text]
 
--- | Helper for getting the token header in the runtime.
-tokenHeaderKey :: IsString s => s
-tokenHeaderKey = fromString $ symbolVal (Proxy @TokenHeaderKey)
+type EntriesAPI = "entries" :> ReqBody '[ JSON] SyncRequest :> PostCreated '[ JSON] NoContent
 
--- | Token for authenticating Hastory Server users.
-newtype Token =
-  Token
-    { unToken :: T.Text
-    }
-  deriving (Show, Eq)
+type ProtectedAPI = Auth '[ JWT] AuthCookie
 
-instance FromHttpApiData Token where
-  parseHeader = fmap Token . parseHeader
-  parseUrlPiece = fmap Token . parseUrlPiece
+type UsersAPI = "users" :> ReqBody '[ JSON] UserForm :> PostCreated '[ JSON] UserId
 
-instance ToHttpApiData Token where
-  toUrlPiece (Token token) = toUrlPiece token
-
-type RequiredHeader = Header' '[ Required, Strict]
+type SessionsAPI
+   = "sessions" :> ReqBody '[ JSON] UserForm :> Verb 'POST 204 '[ JSON] (Headers AuthCookies NoContent)
 
 -- | Main Hastory API specification.
-type HastoryAPI
-   = "commands" :> "append" :> RequiredHeader TokenHeaderKey Token :> ReqBody '[ JSON] SyncRequest :> Post '[ JSON] ()
+type HastoryAPI = UsersAPI :<|> SessionsAPI :<|> (ProtectedAPI :> EntriesAPI)
 
+-- | Proxy for Hastory API.
 api :: Proxy HastoryAPI
 api = Proxy
 
 data HastoryClient =
-  HastoryClient ClientEnv Token
+  HastoryClient
+    { hastoryClientEnv :: ClientEnv
+    , hastoryClientToken :: Token
+    }
+
+-- | An ADT that encodes possible failures for mkHastoryClient
+data ClientEnvFailure
+  = UnableToLogin
+  | NoJWTTokenFound
+  deriving (Show)
 
--- * Hastory Client
 -- | Creates a hastory client type.
 --
--- This type is needed because creating & destroying HTTP managers are expensive.
+-- This type is needed for two reasons. First, because creating and destroying
+-- HTTP managers are expensive. Secondly, the user should only have to log in
+-- successfully once.
+--
 -- Once a user gets a HastoryClient, it's being used throughout the entire life of the user.
-mkHastoryClient :: (MonadError T.Text m, MonadIO m) => BaseUrl -> Token -> m HastoryClient
-mkHastoryClient baseUrl token = do
+mkHastoryClient ::
+     MonadIO m => BaseUrl -> Username -> Text -> m (Either ClientEnvFailure HastoryClient)
+mkHastoryClient url username password = do
   manager <- liftIO $ newManager defaultManagerSettings
-  let clientEnv = mkClientEnv manager baseUrl
-  pure $ HastoryClient clientEnv token
+  let clientEnv = mkClientEnv manager url
+      userForm = mkUserForm (rawUserName username) password
+  res <- liftIO $ runClientM (createSessionClient userForm) clientEnv
+  case res of
+    Left _ -> pure $ Left UnableToLogin
+    Right headers ->
+      case extractJWTCookie headers of
+        Nothing -> pure $ Left NoJWTTokenFound
+        Just token -> pure $ Right (HastoryClient clientEnv token)
+
+-- | Extract token after successful login.
+extractJWTCookie :: MonadError () m => Headers AuthCookies NoContent -> m Token
```

Comment:
MonadError is _really_ weird here. Can we use `Either ClientEnvFailure`
?

Reply:
Done.